qiskit/circuit/register.py

```diff
@@ -117,7 +118,7 @@ def __getitem__(self, key):
             CircuitError: if the `key` is not an integer.
             QiskitIndexError: if the `key` is not in the range `(0, self.size)`.
         """
-        if not isinstance(key, (int, slice, list)):
+        if not isinstance(key, (numbers.Number, int, numbers.Integral, slice, list)):
```

Comment:
It might be a little confusing to allow floats as index. Could we change this to allow only integers and raise an error otherwise? What I meant before was something like 
```suggestion
        if not isinstance(key, (numbers.Integral, slice, list)):
            if isinstance(key, numbers.Number):
                raise CircuitError("register index cannot be a float/complex, must be integer or slice")
```
this way we have a clearer error message than before if the user passed in a float.